### PR TITLE
use abort() to define CAMLunreachable()

### PR DESCRIPTION
**Pull request by gasche**
See https://github.com/ocaml/ocaml/pull/13331#issuecomment-2247673918 for the context.

The new `CAMLunreachable` macro from #13242 has three definitions: one for gcc and clang, one for MSVC, and one for the weird other cases. In the last cases it uses `CAMLassert(0)`, which is clearly wrong for something that should be `[[noreturn]]` (at it does nothing in the non-debug runtime), and will become even more wrong after #13344 (which makes it not-`[[noreturn]]` even in debug mode).

In this version of the definition I propose to simply use `abort()`, which is `[[noreturn]]` as intended and should be available in all C11 compilers.

**Edit**: later iterations of this PR moved to a `caml_abort()` helper, and first some debug printing in DEBUG mode.

cc @MisterDA, @nojb

### CODE REVIEW

**Comment by MisterDA** [comment]
I agree with this, and it was my initial design, but @dustanddreams [had noted](https://github.com/ocaml/ocaml/pull/13242#discussion_r1642736766):

> I'd rather see this defined as [`assert` instead of `abort`] so that, should it ever fire, this gets recognized as an assertion failure rather than a mysterious `SIGABRT`.

and we knew there would be a hole in the definition if we used an assertion. I think a mysterious crash is better than incorrect code, and seeing that LLVM and GCC default to `abort` if they can't provide `__builtin_trap` for a particular architecture makes me think that this is a reasonable work-around.

**Review by MisterDA — APPROVED** [review]

**Comment by ghost** [comment]
I still maintain that, `#ifdef DEBUG`, it would be nice for `CAMLunreachable` to be defined as a form of `CAMLassert`.

**Comment by gasche** [comment]
In #13242, @MisterDA simplified some code in the runtime by *assuming* that the `CAMLunreachable()` macro is `[[noreturn]]`. On the other hand, #13344 goes in the opposite direction by making `CAMLassert(0)` *not*-`[[noreturn]]`. This means that clearly we cannot reuse `CAMLassert(0)` directly. I think that we could share more code between both things -- `caml_failed_assert` and `CAMLunreachable()` could call something like `caml_debug_abort()` that is itself marked `[[noreturn]]`, but this is substantially more work than just tweaking one line as I did. And this work is only relevant for compilers that no one uses, so meh. PRs welcome, but I would suggest going with the one-liner fix I propose in the meantime.

**Comment by gasche** [comment]
(My comment on compilers is not quite right: even on systems that currently define `CAMLunreachable` as `__builtin_trap`, we could reasonably prefer to have some debug output with a source location etc.)

**Comment by gasche** [comment]
What can we do to make progress on this PR (or close it)? @dustanddreams, do you agree that `CAMLunreachable()` should be `[[noreturn]]`, and that `CAMLassert` is not `[[noreturn]]` anymore? Do you think that the best final state is what we currently have in trunk (so: close this PR), or something else than this PR?

**Comment by ghost** [comment]
Well, this PR has an approval so there is nothing getting in its way.

I stated that I would have preferred something like

```
...
#ifdef DEBUG
#define CAMLunreachable() (CAMLassert(0))
#else
#define CAMLunreachable() (abort())
#endif
...
```
but that's only a personal preference. I do not object to the PR as it is.

**Comment by gasche** [comment]
I view things differently:

1. To be mergeable this PR would need a maintainer approval, so I still need to convince someone else to approve the PR. Resolving open discussions in the PR is a reasonable first step before involving someone else.
2. I value your expertise and would not forcefully push a solution that ignores your feedback.

`CAMLassert(0)` will call `caml_failed_assert`, which prints an error message and then calls either `__builtin_trap()` if available and `abort()` otherwise:

https://github.com/ocaml/ocaml/blob/e9374980c47017740f4d6e4981b7579a1f556552/runtime/misc.c#L47-L50

Unless you were thinking of the error message printing (which is rather generic given that `CAMLassert(0)` does not take a string parameter), both solutions appear equivalent to me as they end up calling `abort()`. (This PR only touches the definition of `CAMLunreachable` when `builtin_trap` is not available.) So I think calling `CAMLassert(0)` or `abort()` makes not difference, except that here we do get the `[[noreturn]]` assertion. What am I missing?

**Comment by ghost** [comment]
I think the error message is valuable, even in `CAMLassert(0)`, as it will provide the filename and line number, making it easier to spot which "can't happen" situation actually happened.

File: runtime/misc.c
Comment (ghost) [inline runtime/misc.c:57]: `stderr` is never buffered. It does not make sense to `fflush` it as this is a no-op.

EDIT: I see there are multiple uses of `fflush(stderr)` in OCaml, removing them should be done in a separate PR.

**Comment by gasche** [comment]
I pushed an improved version of this PR to accommodate @dustanddreams' suggestion: if DEBUG is set, `CAMLunreachable()` will print a file and line number on stderr before aborting. The new implementation is more invasive:
- it changes the behavior on all systems (not just those without `__builtin_trap` or `__fastfail`)
- the location-printing code is now shared between `caml_failed_assert` and a new function `caml_debug_abort`

File: runtime/misc.c
Comment (gasche) [reply runtime/misc.c:57]: I thought about this, but the pre-existing code does flush stderr, and people online seem to think that this is still a good idea for portability/compatibility reasons. (Eg. [here](https://utcc.utoronto.ca/~cks/space/blog/programming/HowToWriteToStderr), which also suggests flushing `stdout` before printing to stderr, maybe we should also do this to avoid lost output.)

File: runtime/misc.c
Comment (ghost) [inline runtime/misc.c:40]: This strikes me as a bad idea. If that `strdup` fails for some reason, it will raise an exception, which will either get handled and hide the problem we're attempting to report, or will not but whill hide it anyway...

Comment (gasche) [reply runtime/misc.c:40]: How would you suggest to do things differently? Note that this is exactly the same logic that would run previously on `CAMLassert(0)`.

Comment (ghost) [reply runtime/misc.c:40]: You shuffling things around makes awkward constructs reveal themselves (-:

I don't understand why this argument needs to be `strdup`'ed. It gets immediately passed to `fprintf`. Could this be a leftover from older code which was attempting to do more things with that information?

Comment (gasche) [reply runtime/misc.c:40]: `char_os` is not necessarily the same as `char`, and this function performs the conversion from one to the other.

Comment (ghost) [reply runtime/misc.c:40]: Ah right, damned Windows support...

File: runtime/caml/misc.h
Comment (ghost) [inline runtime/caml/misc.h]: I think it is unnecessary to mention that a routine which prototype is in `misc.h` lives in `misc.c`.
```suggestion
CAMLnoret CAMLextern void caml_debug_abort(char_os *, int);
```

File: runtime/misc.c
Comment (ghost) [inline runtime/misc.c]:
```suggestion
  caml_abort();
```

**Review by ghost — APPROVED** [review]

**Comment by gasche** [comment]
This needs a maintainer approval, and/but both @MisterDA and @dustanddreams reviewed and approved the PR.

**Review by nojb — APPROVED** [review]
Approving on the strength of @dustanddreams and @MisterDA's approvals.

**Comment by gasche** [comment]
Thanks!

File: runtime/misc.c
Comment (NickBarnes) [reply runtime/misc.c:57]: Stdout is often block buffered, and often goes to the same place as stderr, which is why it's always a good idea to flush stdout before writing to stderr. I _think_ that (Posix) stderr is never block buffered (unless you `setvbuf` it?) but may be line-buffered, rather than unbuffered (I think on Linux it's always unbuffered); so it may be a good idea to flush it after writing a part-line, and after any write if it may have been `setvbuf`ed.

**Pull request merged**
